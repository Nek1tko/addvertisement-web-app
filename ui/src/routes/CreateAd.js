import React, { useState, useEffect } from "react";
import { Box, MenuItem, InputAdornment } from "@material-ui/core";
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import LocalizedDropzoneArea from '../components/LocalizedDropzoneArea';
import { Collapse } from '@material-ui/core';
import { Alert } from "@mui/material";
import AuthService from '../services/auth.service';
import { Redirect } from "react-router-dom";
import axios from 'axios';
import authHeader from "../services/auth-header";

const API_URL = "http://localhost:8080/";

const CreateAdImpl = props => {
    const { history } = props;

    const [name, setName] = useState('');
    const [description, setDescription] = useState('');
    const [price, setPrice] = useState('');

    const [metroList, setMetroList] = useState([]);
    const [metro, setMetro] = useState('');

    const [categoryList, setCategoryList] = useState([]);
    const [category, setCategory] = useState('');

    const [subcategoryList, setSubcategoryList] = useState([]);
    const [subcategory, setSubcategory] = useState('');

    const [alertOpen, setAlertOpen] = useState(false);
    const [errorMessage, setErrorMessage] = useState('');

    const { _, userId, } = AuthService.getUser();

    useEffect(() => {
        axios
            .get(API_URL + "metro", { headers: authHeader() })
            .then(res => {
                setMetroList(res.data);
            })
    }, []);

    useEffect(() => {
        axios
            .get(API_URL + "category", { headers: authHeader() })
            .then(res => {
                setCategoryList(res.data);
            })
    }, []);

    useEffect(() => {
        axios
            .get(API_URL + "category/" + category, { headers: authHeader() })
            .then(res => {
                setSubcategoryList(res.data);
                setSubcategory('');
            })
    }, [category]);

    const handleSubmit = e => {
        e.preventDefault();
        if (name.length < 3 || name.length > 50) {
            setErrorMessage("Название объявления должно быть длиной от 3 до 50 символов");
            setAlertOpen(true);
            return;
        }

        axios
            .post(API_URL + "ad", {
                name: name,
                description: description,
                price: price,
                saler: {
                    id: userId
                },
                metro: {
                    id: metro
                },
                subCategory: {
                    id: subcategory
                },
                isActive: true
            }, { headers: authHeader() })
            .then(res => {
                history.push('/my-ads');
            })

        history.push('/my-ads');
    };

    return (
        <Box
            display='flex'
            flexDirection='column'
            justifyContent='space-between'
            sx={{ mt: 2, width: '50%', m: 'auto' }}
        >
            <Collapse in={alertOpen}>
                <Alert
                    severity="error"
                    sx={{ mb: 2 }}
                >
                    {errorMessage}
                </Alert>
            </Collapse>

            <TextField
                label="Название"
                variant="filled"
                fullWidth
                onChange={e => {
                    setName(e.target.value);
                    setAlertOpen(false);
                }}
                value={name}
            />

            <TextField
                label="Цена"
                variant="filled"
                type="number"
                fullWidth
                InputProps={{
                    endAdornment: <InputAdornment position="end">₽</InputAdornment>,
                }}
                onChange={e => {
                    if (e.target.value >= 0) {
                        setPrice(e.target.value);
                    }
                }}
                value={price}
            />

            <TextField
                label="Описание"
                variant="filled"
                multiline
                maxRows={8}
                minRows={8}
                onChange={e => {
                    setDescription(e.target.value);
                    setAlertOpen(false);
                }}
                value={description}
                fullWidth
            />

            <TextField
                select
                variant="filled"
                label="Метро"
                value={metro}
                onChange={e => {
                    setMetro(e.target.value);
                }}
            >
                {metroList.map(metro => {
                    return (
                        <MenuItem key={metro.id} value={metro.id}>
                            {metro.name}
                        </MenuItem>);
                })}
            </TextField>

            <TextField
                select
                variant="filled"
                label="Категория"
                value={category}
                onChange={e => {
                    const value = e.target.value;
                    setCategory(value);
                    setSubcategory('');
                }}
            >
                {categoryList.map(category => {
                    return (
                        <MenuItem
                            key={category.id}
                            value={category.id}
                        >
                            {category.name}
                        </MenuItem>);
                })}
            </TextField>

            <TextField
                disabled={!category}
                select
                variant="filled"
                label="Подкатегория"
                value={subcategory}
                onChange={e => {
                    const value = e.target.value;
                    setSubcategory(value);
                }}
            >
                {subcategoryList.map(subcategory => {
                    return (
                        <MenuItem
                            key={subcategory.id}
                            value={subcategory.id}
                        >
                            {subcategory.name}
                        </MenuItem>);
                })}
            </TextField>

            <LocalizedDropzoneArea
                filesLimit={3}
                maxFileSize={3145728} // 3 mb
                acceptedFiles={[".jpeg", ".jpg"]}
            />

            <Button
                type="submit"
                variant="contained"
                color="primary"
                onClick={handleSubmit}
                fullWidth
                style={{ marginTop: 20 }}
                disabled={!name || !description || !price || !metro || !category || !subcategory}
            >
                Создать
            </Button>
        </Box >
    )
}

const CreateAd = props => {
    if (!AuthService.getUser()) {
        return <Redirect to="/login" />
    }

    return <CreateAdImpl history={props.history} />
};

export default CreateAd;
