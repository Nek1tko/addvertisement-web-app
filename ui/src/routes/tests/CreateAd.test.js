/**
 * @group unit
 */

import axios from "axios";
import { mount } from "enzyme";
import CreateAd from "../CreateAd";
import { act } from "react-dom/test-utils";
import { Router } from "react-router-dom";
import { createBrowserHistory } from 'history';
import React from "react";
import AuthService from "../../services/auth.service";
import authHeader from "../../services/auth-header";

jest.mock('form-data');
jest.mock('axios');
jest.mock("../../services/auth.service");
jest.mock("../../services/auth-header");


const event = {
    preventDefault: jest.fn()
};

const eventName = {
    target: {
        value: "mockName"
    }
};

const eventName2 = {
    target: {
        value: "m"
    }
};

const eventDescription = {
    target: {
        value: "mockDescription"
    }
};

const eventPrice = {
    target: {
        value: 156
    }
};

const eventCategory = {
    target: {
        value: "mockCategory"
    }
};

const eventSubCategory = {
    target: {
        value: "mockSubCategory"
    }
};

const eventMetro = {
    target: {
        value: "mockMetro"
    }
};

const imgs = [
    {
        name: JSON.stringify({ file: 'mockImg' })
    }
];


describe('CreateAdTests', () => {
    let wrapper, props;
    const setState = jest.fn();

    beforeEach(() => {
        props = {
            history: {
                ...createBrowserHistory(),
                push: jest.fn()
            }
        };

        FormData = jest.fn();
        AuthService.getUser.mockReturnValue({ user: "mockUser" });
        authHeader.mockReturnValue({ header: "mockHeader" });
        axios.get.mockResolvedValue({
            data: [
                {
                    id: 1,
                    name: "mock name"
                },
                {
                    id: 2,
                    name: "mock name"
                },
            ]
        });
        axios.post.mockResolvedValue({
            data: {
                id: "ID"
            }
        });
        const useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation((state) => [state, setState]);
        wrapper = mount(
            <Router history={props.history}>
                <CreateAd {...props} />
            </Router>);
    })

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('CreateAdRenderTest', () => {
        act(() => {
            mount(<Router history={props.history}>
                <CreateAd {...props} />
            </Router>);
        });
    });

    test('CreateAdSnapshotTest', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('CreateAdSubmitAdPositiveTest', async () => {
        //setup all necessary fields
        await act(async () => {
            await wrapper.find('#adNameTextField').first().props().onChange(eventName);
            await wrapper.find('#categoryTextField').first().props().onChange(eventCategory);
            await wrapper.find('#subcategoryTextField').first().props().onChange(eventSubCategory);
            await wrapper.find('#descriptionTextField').first().props().onChange(eventDescription);
            await wrapper.find('#priceTextField').first().props().onChange(eventPrice);
            await wrapper.find('#metroTextField').first().props().onChange(eventMetro);
            await wrapper.find('#localizedDropzoneArea').props().onChange(imgs);

            await wrapper.update();
        });

        await act(async () => {
            await wrapper.find('#submitButton').first().props().onClick(event);
        });

        expect(props.history.push).toHaveBeenCalledTimes(1);
        expect(props.history.push).toHaveBeenCalledWith("/my-ads");
    });

    test('CreateAdSubmitAdNegativeTest', async () => {
        //setup all necessary fields
        await act(async () => {
            await wrapper.find('#adNameTextField').first().props().onChange(eventName2);
            await wrapper.find('#categoryTextField').first().props().onChange(eventCategory);
            await wrapper.find('#subcategoryTextField').first().props().onChange(eventSubCategory);
            await wrapper.find('#descriptionTextField').first().props().onChange(eventDescription);
            await wrapper.find('#priceTextField').first().props().onChange(eventPrice);
            await wrapper.find('#metroTextField').first().props().onChange(eventMetro);
            await wrapper.find('#localizedDropzoneArea').props().onChange(imgs);

            await wrapper.update();
        });

        await act(async () => {
            await wrapper.find('#submitButton').first().props().onClick(event);
        });

        expect(props.history.push).toHaveBeenCalledTimes(0);
    });

    test('CreateAdNotAuthorizedTest', () => {
        AuthService.getUser.mockReturnValue(null);
        wrapper = mount(
            <Router history={props.history}>
                <CreateAd {...props} />
            </Router>);
    });
});
