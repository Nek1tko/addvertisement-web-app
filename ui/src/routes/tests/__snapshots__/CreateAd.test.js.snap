// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAdTests CreateAdSnapshotTest 1`] = `ReactWrapper {}`;

exports[`CreateAdTests SellerAdSnapshotTest 1`] = `ReactWrapper {}`;
